builtin/logical/ssh/util.go

```diff
@@ -0,0 +1,206 @@
+	}
+	defer session.Close()
+	go func() {
+		w, _ := session.StdinPipe()
+		fmt.Fprintln(w, "C0644", len(publicKey), dynamicPublicKeyFileName)
+		io.Copy(w, strings.NewReader(publicKey))
+		fmt.Fprint(w, "\x00")
+		w.Close()
+	}()
+	err = session.Run(fmt.Sprintf("scp -vt %s", dynamicPublicKeyFileName))
+	if err != nil {
+		return fmt.Errorf("public key upload failed")
+	}
+	return nil
+}
+
+// Creates a SSH session object which can be used to run commands
+// in the target machine. The session will use public key authentication
+// method with port 22.
+func createSSHPublicKeysSession(username, ipAddr, port, hostKey string) (*ssh.Session, error) {
+	if username == "" {
+		return nil, fmt.Errorf("missing username")
+	}
+	if ipAddr == "" {
+		return nil, fmt.Errorf("missing ip address")
+	}
+	if hostKey == "" {
+		return nil, fmt.Errorf("missing host key")
+	}
+	signer, err := ssh.ParsePrivateKey([]byte(hostKey))
+	if err != nil {
+		return nil, fmt.Errorf("parsing Private Key failed: %s", err)
+	}
+
+	config := &ssh.ClientConfig{
+		User: username,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+	}
+
+	client, err := ssh.Dial("tcp", fmt.Sprintf("%s:%s", ipAddr, port), config)
+	if err != nil {
+		return nil, err
+	}
+	if client == nil {
+		return nil, fmt.Errorf("invalid client object: %s", err)
+	}
+
+	session, err := client.NewSession()
+	if err != nil {
+		return nil, err
+	}
+	return session, nil
+}
+
+// Creates a new RSA key pair with key length of 2048.
+// The private key will be of pem format and the public key will be
+// of OpenSSH format.
+func generateRSAKeys() (publicKeyRsa string, privateKeyRsa string, err error) {
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return "", "", fmt.Errorf("error generating RSA key-pair: %s", err)
+	}
+
+	privateKeyRsa = string(pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+	}))
+
+	sshPublicKey, err := ssh.NewPublicKey(privateKey.Public())
+	if err != nil {
+		return "", "", fmt.Errorf("error generating RSA key-pair: %s", err)
+	}
+	publicKeyRsa = "ssh-rsa " + base64.StdEncoding.EncodeToString(sshPublicKey.Marshal())
+	return
+}
+
+// Concatenates the public present in that target machine's home
+// folder to ~/.ssh/authorized_keys file
+func installPublicKeyInTarget(adminUser, username, ip, port, hostKey string) error {
+	session, err := createSSHPublicKeysSession(adminUser, ip, port, hostKey)
+	if err != nil {
+		return fmt.Errorf("unable to create SSH Session using public keys: %s", err)
+	}
+	if session == nil {
+		return fmt.Errorf("invalid session object")
+	}
+	defer session.Close()
+
+	authKeysFileName := fmt.Sprintf("/home/%s/.ssh/authorized_keys", username)
+	tempKeysFileName := fmt.Sprintf("/home/%s/temp_authorized_keys", username)
+
+	// Commands to be run on target machine
+	dynamicPublicKeyFileName := fmt.Sprintf("vault_ssh_%s_%s.pub", username, ip)
```

Comment:
This process seem brittle to me. It may be better to provide a simple shell script which they drop in /usr/local/bin and configure the roll with the command like "/usr/local/bin/install_key.sh" and we provide that in our documentation. This worries me...